pkg/innerring/audit.go

```diff
@@ -0,0 +1,193 @@
+	return (*containerAPI.Container)(c).OwnerID()
+}
+
+func (a auditSettlementDeps) AuditResultsForEpoch(epoch uint64) ([]*auditAPI.Result, error) {
+	idList, err := a.auditClient.ListAuditResultIDByEpoch(epoch)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list audit results in sidechain")
+	}
+
+	res := make([]*auditAPI.Result, 0, len(idList))
+
+	for i := range idList {
+		r, err := a.auditClient.GetAuditResult(idList[i])
+		if err != nil {
+			return nil, errors.Wrap(err, "could not get audit result")
+		}
+
+		res = append(res, r)
+	}
+
+	return res, nil
+}
+
+func (a auditSettlementDeps) ContainerInfo(cid *containerAPI.ID) (audit.ContainerInfo, error) {
+	cnr, err := a.cnrSrc.Get(cid)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get container from storage")
+	}
+
+	return (*containerWrapper)(cnr), nil
+}
+
+func (a auditSettlementDeps) buildContainer(e uint64, cid *containerAPI.ID) (netmapAPI.ContainerNodes, *netmapAPI.Netmap, error) {
+	var (
+		nm  *netmapAPI.Netmap
+		err error
+	)
+
+	if e > 0 {
+		nm, err = a.nmSrc.GetNetMapByEpoch(e)
+	} else {
+		nm, err = netmap.GetLatestNetworkMap(a.nmSrc)
+	}
+
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not get network map from storage")
+	}
+
+	cnr, err := a.cnrSrc.Get(cid)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not get container from sidechain")
+	}
+
+	cn, err := nm.GetContainerNodes(
+		cnr.PlacementPolicy(),
+		cid.ToV2().GetValue(), // may be replace pivot calculation to neofs-api-go
+	)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "could not calculate container nodes")
+	}
+
+	return cn, nm, nil
+}
+
+func (a auditSettlementDeps) ContainerNodes(e uint64, cid *containerAPI.ID) ([]audit.NodeInfo, error) {
+	cn, _, err := a.buildContainer(e, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := cn.Flatten()
+	res := make([]audit.NodeInfo, 0, len(ns))
+
+	for i := range ns {
+		res = append(res, &nodeInfoWrapper{
+			ni: ns[i],
+		})
+	}
+
+	return res, nil
+}
+
+func (a auditSettlementDeps) SGInfo(addr *object.Address) (audit.SGInfo, error) {
+	cn, nm, err := a.buildContainer(0, addr.ContainerID())
+	if err != nil {
+		return nil, err
+	}
+
+	sg, err := a.clientCache.getSG(context.Background(), addr, nm, cn)
+	if err != nil {
+		return nil, err
+	}
+
+	return (*sgWrapper)(sg), nil
+}
+
+func (a auditSettlementDeps) ResolveKey(ni audit.NodeInfo) (*owner.ID, error) {
+	w, err := owner.NEO3WalletFromPublicKey(crypto.UnmarshalPublicKey(ni.PublicKey()))
+	if err != nil {
+		return nil, err
+	}
+
+	id := owner.NewID()
+	id.SetNeo3Wallet(w)
+
+	return id, nil
+}
+
+var transferAuditDetails = []byte("settlement-audit")
+
+func (a auditSettlementDeps) Transfer(sender, recipient *owner.ID, amount *big.Int) {
+	if !amount.IsInt64() {
+		a.log.Error("amount can not be represented as an int64",
+			zap.Stringer("value", amount),
+		)
+
+		return
+	}
+
+	amount64 := amount.Int64()
+	if amount64 == 0 {
```

Comment:
I guess this condition is used to transfer at least 1 molecule of GAS. Not sure if it is correct place for this condition.

Consider there are `a->b:100` and `b->a:100` transactions. Both transfer 100 molecules of GAS so no extra molecules required. After adding these transactions to table, it will contain single record `a->b:0` and we should not transfer any assets from one to another.

According to this code we will transfer 1 molecule from `a` to `b`, but we should not.

Instead this branch should be used in `fillTransferTable()`:
```go
		fee := big.NewInt(0).Mul(price, ctx.sumSGSize)
		fee.Div(fee, bigGB)

		if fee.Int64() == 0 {
			fee = big.NewInt(1)
		}
```

Reply:
Fixed